src/v/cloud_storage/cache_service.cc

```diff
@@ -57,6 +58,68 @@ cache::cache(
   , _cnt(0)
   , _total_cleaned(0) {}
 
+ss::future<>
+cache::recursive_delete_empty_directory(const std::string_view& key) {
+    gate_guard guard{_gate};
+
+    std::filesystem::path canonical_path;
+    try {
+        canonical_path = std::filesystem::canonical(key);
+    } catch (std::filesystem::filesystem_error& e) {
+        if (e.code() == std::errc::no_such_file_or_directory) {
+            vlog(cst_log.debug, "Tried to remove {} that doesn't exist", key);
+            co_return;
+        } else {
+            throw;
+        }
+    }
+
+    std::filesystem::path canonical_cache_dir;
+    try {
+        canonical_cache_dir = std::filesystem::canonical(_cache_dir);
+    } catch (std::filesystem::filesystem_error& e) {
+        if (e.code() == std::errc::no_such_file_or_directory) {
+            vlog(
+              cst_log.debug,
+              "Tried to remove {}, but _cache_dir {} doesn't exist",
+              key,
+              _cache_dir);
+            co_return;
+        } else {
+            throw;
+        }
+    }
+
+    if (!std::string_view(canonical_path.native())
+           .starts_with(canonical_cache_dir.native())) {
+        throw std::invalid_argument(fmt_with_ctx(
+          fmt::format,
+          "Tried to clean up {}, which is outside of cache_dir {}.",
+          canonical_path.native(),
+          canonical_cache_dir.native()));
+    }
+
+    while (canonical_path != canonical_cache_dir) {
```

Comment:
what happens if canonical_path was not a child of canonical_cache_dir to begin with?

Reply:
then `invalid_argument` is thrown on line 95